diff -urNp coreutils-8.4-orig/m4/xattr.m4 coreutils-8.4/m4/xattr.m4
--- coreutils-8.4-orig/m4/xattr.m4	2010-01-12 07:36:58.000000000 +0100
+++ coreutils-8.4/m4/xattr.m4	2010-06-14 09:40:44.922293550 +0200
@@ -33,11 +33,11 @@ AC_DEFUN([gl_FUNC_XATTR],
         use_xattr=yes
       fi
     fi
-    AC_DEFINE_UNQUOTED([USE_XATTR], [$use_xattr],
-                       [Define if you want extended attribute support.])
     if test $use_xattr = no; then
       AC_MSG_WARN([libattr development library was not found or not usable.])
       AC_MSG_WARN([AC_PACKAGE_NAME will be built without xattr support.])
     fi
   fi
+  AC_DEFINE_UNQUOTED([USE_XATTR], [`test $use_xattr != yes; echo $?`],
+                     [Define if you want extended attribute support.])
 ])
