--- sh-utils-1.16/src/su.c.badpaths	Mon Apr 14 14:26:55 1997
+++ sh-utils-1.16/src/su.c	Sun Aug 17 14:11:31 EDT 2003
@@ -147,6 +147,15 @@
 # define DEFAULT_ROOT_LOGIN_PATH "/usr/ucb:/bin:/usr/bin:/etc"
 #endif
 
+/* The default paths which get set are both bogus and oddly influenced
+   by <paths.h> and -D on the commands line. Just to be clear, we'll set
+   these explicitly. -ewt */
+#undef DEFAULT_LOGIN_PATH
+#undef DEFAULT_ROOT_LOGIN_PATH
+#define DEFAULT_LOGIN_PATH "/usr/local/bin:/bin:/usr/bin"
+#define DEFAULT_ROOT_LOGIN_PATH \
+    "/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin"
+
 /* The shell to run if none is given in the user's passwd entry.  */
 #define DEFAULT_SHELL "/bin/sh"
 
